src/components/CytoscapeGraph/graphs/TrafficRenderer.ts

```diff
@@ -0,0 +1,292 @@
+  private points: Array<TrafficPoint> = [];
+  private generator: TrafficPointGenerator;
+  private edge: any;
+
+  constructor(speed: number, timer: number | undefined, edge: any) {
+    this.generator = new TrafficPointGenerator(speed, timer);
+    this.edge = edge;
+  }
+
+  /**
+   * Process a step for the Traffic Edge, increments the delta of the points
+   * by step / speed and clamps to (-inf, 1)
+   * In charge of calling the `processStep` for the generator and adding a new point if any.
+   */
+  processStep(step: number) {
+    this.points = this.points.map(p => {
+      p.delta += step / p.speed;
+      p.delta = Math.min(p.delta, 1);
+      return p;
+    });
+    const point = this.generator.processStep(step);
+    if (point) {
+      this.points.push(point);
+    }
+  }
+
+  getPoints() {
+    return this.points;
+  }
+
+  getEdge() {
+    return this.edge;
+  }
+
+  setTimer(timer: number | undefined) {
+    this.generator.setTimer(timer);
+  }
+
+  /**
+   * When a point is 1 or over it, is time to discard it.
+   */
+  removeFinishedPoints() {
+    this.points = this.points.filter(p => p.delta < 1);
+  }
+
+  setSpeed(speed: number) {
+    this.generator.setSpeed(speed);
+  }
+
+  setEdge(edge: any) {
+    this.edge = edge;
+  }
+}
+
+type TrafficEdgeHash = {
+  [edgeId: string]: TrafficEdge;
+};
+
+let test = 0;
+
+/**
+ * Renders the traffic going from edges using the edge information to compute
+ * their rate and speed
+ */
+export default class TrafficRenderer {
+  private animationTimer;
+  private previousTimestamp;
+  private trafficEdges: TrafficEdgeHash = {};
+
+  private layer;
+  private canvas;
+  private context;
+  private test;
+
+  constructor(cy: any, edges: any) {
+    this.layer = cy.cyCanvas();
+    this.canvas = this.layer.getCanvas();
+    this.context = this.canvas.getContext('2d');
+    this.setEdges(edges);
+    this.test = test++;
+  }
+
+  /**
+   * Starts the rendering loop, discards any other rendering loop that was started
+   */
+  start() {
+    this.stop();
+    console.log('Remove: Starting animation');
+    this.animationTimer = window.setInterval(this.processStep, FRAME_RATE * 1000);
+  }
+
+  /**
+   * Stops the rendering loop if any
+   */
+  stop() {
+    console.log('Remove: Might stop the animation (if any)');
+    if (this.animationTimer) {
+      console.log('Remove: Stopping the animation');
+      window.clearInterval(this.animationTimer);
+      this.animationTimer = undefined;
+      this.clear();
+    }
+  }
+
+  setEdges(edges: any) {
+    this.trafficEdges = this.processEdges(edges);
+  }
+
+  clear() {
+    this.layer.clear(this.context);
+  }
+
+  /**
+   * Process a step, clears the canvas, sets the graph transformation to render
+   * every dot.
+   */
+  processStep = () => {
+    console.log('Hello from Renderer:', this.test);
+    if (this.previousTimestamp === undefined) {
+      this.previousTimestamp = Date.now();
+    }
+    const nextTimestamp = Date.now();
```

Comment:
Very minor thing: `Date.now()` is executed here and the line below in `this.currentStep()` call. It could be passed as parameter to `currentStep` instead. Also, I wonder if the very first `step` could be 0 and if that could lead to some "div by 0" bug ? To prevent that you could initialize the first step to `FRAME_RATE * 1000`

Reply:
I don't recall any div by currentStep, but will take this into consideration, thanks.